discretize/curvilinear_mesh.py

```diff
@@ -0,0 +1,437 @@
+
+
+def _normalize2D(x):
+    return x / np.kron(np.ones((1, 2)), mkvc(_length2D(x), 2))
+
+
+def _normalize3D(x):
+    return x / np.kron(np.ones((1, 3)), mkvc(_length3D(x), 2))
+
+
+class CurvilinearMesh(BaseRectangularMesh, DiffOperators, InnerProducts):
+    """CurvilinearMesh is a mesh class that deals with curvilinear meshes.
+
+    Example of a curvilinear mesh:
+
+    .. plot::
+        :include-source:
+
+        import discretize
+        X, Y = discretize.utils.exampleLrmGrid([3,3],'rotate')
+        mesh = discretize.CurvilinearMesh([X, Y])
+        mesh.plot_grid(show_it=True)
+    """
+
+    _meshType = "Curv"
+    _aliases = {
+        **DiffOperators._aliases,
+        **BaseRectangularMesh._aliases,
+        **{
+            "gridCC": "grid_cell_centers",
+            "gridN": "grid_nodes",
+            "gridFx": "grid_faces_x",
+            "gridFy": "grid_faces_y",
+            "gridFz": "grid_faces_z",
+            "gridEx": "grid_edges_x",
+            "gridEy": "grid_edges_y",
+            "gridEz": "grid_edges_z",
+        },
+    }
+
+    nodes = properties.List(
+        "List of arrays describing the node locations",
+        prop=properties.Array(
+            "node locations in an n-dimensional array",
+            shape={("*", "*"), ("*", "*", "*")},
+        ),
+        min_length=2,
+        max_length=3,
+    )
+
+    def __init__(self, nodes=None, **kwargs):
+
+        self.nodes = nodes
+
+        if "_n" in kwargs.keys():
+            n = kwargs.pop("_n")
+            if np.any(n != np.array(self.nodes[0].shape) - 1):
+                raise ValueError(
+                    "Unexpected n-values. {} was provided, {} was expected".format(
+                        n, np.array(self.nodes[0].shape) - 1
+                    )
+                )
+        else:
+            n = np.array(self.nodes[0].shape) - 1
+
+        BaseRectangularMesh.__init__(self, n, **kwargs)
+
+        # Save nodes to private variable _gridN as vectors
+        self._gridN = np.ones((self.nodes[0].size, self.dim))
+        for i, node_i in enumerate(self.nodes):
+            self._gridN[:, i] = mkvc(node_i.astype(float))
+
+    @properties.validator("nodes")
+    def _check_nodes(self, change):
+        if len(change["value"]) <= 1:
+            raise ValueError("len(node) must be greater than 1")
+
+        for i, change["value"][i] in enumerate(change["value"]):
+            if change["value"][i].shape != change["value"][0].shape:
+                raise ValueError(
+                    "change['value'][{0:d}] is not the same shape as "
+                    "change['value'][0]".format(i)
+                )
+
+        if len(change["value"][0].shape) != len(change["value"]):
+            raise ValueError("Dimension mismatch")
+
+    @property
+    def grid_cell_centers(self):
+        """
+        Cell-centered grid
+        """
+        if getattr(self, "_gridCC", None) is None:
+            self._gridCC = np.concatenate(
```

Comment:
to simplify maintenance down the line, I think having the private variable named the same as the public, just with the underscore would be helpful `self._grid_cell_centers`. Do you think there are any things we should be concerned about for doing that at this stage?